lib/common-types.ts

```diff
@@ -1,13 +1,48 @@
 import { notStrictEqual } from 'assert'
 
+/**
+ * An object whose all properties have the same type.
+ */
 export type Dictionary<T = any> = { [key: string]: T }
 
+/**
+ * Returns the keys of T that match Dictionary<U> and are not arrays.
+ */
+export type DictionaryKeyof<T, U = any> = Exclude<KeyOf<T, Dictionary<U>>, KeyOf<T, any[]>>
+
+/**
+ * Returns the keys of T that match U.
+ */
+export type KeyOf<T, U> = Exclude<{ [K in keyof T]: T[K] extends U ? K : never }[keyof T], undefined>
+
+/**
+ * An array whose first element is not undefined.
+ */
 export type NotEmptyArray<T = any> = [T, ...T[]]
 
-export function assertNotTrue<T = any> (actual: T | true): asserts actual is T {
-  notStrictEqual(actual, true)
+/**
+ * Returns the type of a Dictionary or array values.
+ */
+export type ValueOf<T> = T extends (infer U)[] ? U : T[keyof T];
+
+/**
+ * Typing wraper around assert.notStrictEqual()
+ */
+export function assertNotStrictEqual<N, T> (actual: T|N, expected: N, message ?: string | Error)
+: asserts actual is Exclude<T, N> {
+  notStrictEqual(actual, expected, message)
+}
+
+/**
+ * Asserts actual is neither a Dictionary nor an array.
+ */
+export function assertSingleKey (actual: string | string[] | Dictionary): asserts actual is string {
+  notStrictEqual(typeof actual, 'object')
```

Comment:
I'd missed this line in review, we actually have a helper library called `argsert`, the goal of which being to assert folks are passing in the correct argument types, and enforce it. I suggest we use this consistently.

argsert grew out of folks often passing an object into a method which took a string, and then opening issues on yargs.

Reply:
Fixed.

> we actually have a helper library called argsert, the goal of which being to assert folks are passing in the correct argument types, and enforce it. I suggest we use this consistently.

Yes, but the 2 problem I ran into with argsert are:
- as it takes only a single signature, when most functions in yargs do have multiple ones, it usually accepts a too broad set of inputs
- as the signature is defined by a string, it is not possible to infer type assertions from it

Hence the need for these complementary assert functions.

Reply:
A general refactor of argsert (merging argsert and the new assertXxx functions) for consistency may be something worth doing. But it would probably mean replacing the single argsert function by several argsertXxx functions, as the argsert way of describing signature is not type inference compatible.